test/functorch/test_control_flow.py

```diff
@@ -1471,44 +1726,860 @@ def fct_pointwise(x, y):
+        result_init = scan_fct(op, input=inp, dim=dim, reverse=reverse, init=init)
+        result_exp = _fake_scan(op, init=init, input=inp, dim=dim, reverse=reverse)
+        self.assertEqual(result_init, result_exp)
+        self.assertEqual(result_init[0].shape, torch.Size([2, 5, 2]))
+
+        # Init wrong shape (Other dim different)
+        init = torch._ops.ops.aten.slice(x, dim, 0, 1, 1)
+        init = torch.tile(init, (1, 2, 1))
+        with self.assertRaisesRegex(
+            RuntimeError,
+            "The size of tensor a.*",
+        ):
+            result_init = scan_fct(op, input=inp, dim=dim, reverse=reverse, init=init)
+
+        def add_scalar_carry_sliced_out(x: torch.Tensor, y: torch.Tensor):
+            return x + 1.0, x[:, :1, :] + y
+
+        result_init = scan_fct(
+            add_scalar_carry_sliced_out, input=inp, dim=dim, reverse=reverse, init=init
+        )
+        result_exp = _fake_scan(
+            add_scalar_carry_sliced_out, init=init, input=inp, dim=dim, reverse=reverse
+        )
+        self.assertEqual(result_init, result_exp)
+        self.assertEqual(result_init[0].shape, torch.Size([1, 10, 2]))
+        self.assertEqual(result_init[1].shape, inp.shape)
+
+        # Init wrong pytree
+        init = (
+            torch._ops.ops.aten.slice(x, dim, 0, 1, 1),
+            torch._ops.ops.aten.slice(x, dim, 0, 1, 1),
+        )
+
+        def add_one_carry(x: torch.Tensor, y: torch.Tensor):
+            return x[0], x
+
+        with self.assertRaisesRegex(
+            # Should be: RuntimeError: The number of leaves of the pytree of the new carry produced
+            # by the operator needs to match the length of the pytree of the init
+            torch._dynamo.exc.Unsupported,
+            "Observed exception.*",
+        ):
+            result_init = scan_fct(
+                add_one_carry, input=inp, dim=dim, reverse=reverse, init=init
+            )
+
+        # Correct case
+        op, op_pt = (add, torch.cumsum)
+        x = torch.randn(3, 2, 2, device=device)
+        dim = 1
+
+        if reverse:
+            init = torch.zeros_like(torch._ops.ops.aten.slice(x, dim, -1, None, 1))
+            inp = torch._ops.ops.aten.slice(x, dim, 0, -1, 1)
+        else:
+            init = torch.zeros_like(torch._ops.ops.aten.slice(x, dim, 0, 1, 1))
+            inp = torch._ops.ops.aten.slice(x, dim, 1, None, 1)
+
+        result = scan_fct(op, init=init, input=x, dim=dim, reverse=reverse)
+        result_exp = _fake_scan(op, init=init, input=x, dim=dim, reverse=reverse)
+
+        self.assertEqual(result, result_exp)
+        if not reverse:
+            result_exp_PT = op_pt(x, dim)
+            self.assertEqual(result[1], result_exp_PT)
+
+    @requires_cuda
+    @parametrize("reverse", [False, True])
+    @parametrize("device", [torch.device("cpu"), torch.device("cuda")])
+    def test_scan_init_complex_pytree(self, reverse, device):
+        def fct_pointwise_carry_wrong_pytree(x, y):
```

Comment:
We should probably have a collection of combine_fn and parameterize over them so that they can be shared among tests.

Reply:
I added a [collection](https://github.com/pytorch/pytorch/blob/cb06c9cf487b748caa807bcc66207ce2f57f55cf/test/functorch/test_control_flow.py#L167-L223) of such function and consolidated their usecases in the tests